File: owncloudApp/src/main/java/com/owncloud/android/ui/activity/FolderPickerActivity.java
/**
 * ownCloud Android client application
 *
 * @author Shashvat Kedia
 * @author David González Verdugo
 * @author Abel García de Prada
 * Copyright (C) 2020 ownCloud GmbH.
 * <p>
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2,
 * as published by the Free Software Foundation.
 * <p>
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * <p>
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.owncloud.android.ui.activity;

import android.accounts.Account;
import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import com.owncloud.android.R;
import com.owncloud.android.datamodel.FileDataStorageManager;
import com.owncloud.android.domain.files.model.FileListOption;
import com.owncloud.android.domain.files.model.OCFile;
import com.owncloud.android.domain.spaces.model.OCSpace;
import com.owncloud.android.presentation.files.filelist.MainFileListFragment;
import com.owncloud.android.ui.fragment.FileFragment;
import com.owncloud.android.utils.PreferenceUtils;
import timber.log.Timber;

import java.util.ArrayList;

public class FolderPickerActivity extends FileActivity implements FileFragment.ContainerActivity,
        OnClickListener, MainFileListFragment.FileActions {

    public static final String EXTRA_FOLDER = FolderPickerActivity.class.getCanonicalName()
            + ".EXTRA_FOLDER";
    public static final String EXTRA_FILES = FolderPickerActivity.class.getCanonicalName()
            + ".EXTRA_FILES";

    private static final String TAG_LIST_OF_FOLDERS = "LIST_OF_FOLDERS";

    public static final String EXTRA_PICKER_OPTION = "EXTRA_PICKER_OPTION";

    protected Button mCancelBtn;
    protected Button mChooseBtn;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Timber.d("onCreate() start");

        super.onCreate(savedInstanceState);

        setContentView(R.layout.files_folder_picker);     // beware - inflated in other activities too

        // Allow or disallow touches with other visible windows
        LinearLayout filesFolderPickerLayout = findViewById(R.id.filesFolderPickerLayout);
        filesFolderPickerLayout.setFilterTouchesWhenObscured(
                PreferenceUtils.shouldDisallowTouchesWithOtherVisibleWindows(this)
        );

        if (savedInstanceState == null) {
            initAndShowListOfFilesFragment();
        }

        // sets callback listeners for UI elements
        initPickerListeners();

        // Action bar setup
        setupStandardToolbar(null, false, false, true);

        // Set action button text
        setActionButtonText();

        Timber.d("onCreate() end");
    }

    /**
     * Called when the ownCloud {@link Account} associated to the Activity was just updated.
     */
    @Override
    protected void onAccountSet(boolean stateWasRecovered) {
        super.onAccountSet(stateWasRecovered);
        if (getAccount() != null) {

            updateFileFromDB();

            OCFile folder = getFile();
            if (folder == null || !folder.isFolder()) {
                // fall back to root folder
                setFile(getStorageManager().getFileByPath(OCFile.ROOT_PATH));
                folder = getFile();
            }

            if (!stateWasRecovered) {
                MainFileListFragment listOfFolders = getListOfFilesFragment();
                listOfFolders.navigateToFolder(folder);
            }

            updateNavigationElementsInActionBar();
        }
    }

    private void initAndShowListOfFilesFragment() {
        OCFile safeInitialFolder;
        if (getFile() == null) {
            FileDataStorageManager fileDataStorageManager = new FileDataStorageManager(getAccount());
            safeInitialFolder = fileDataStorageManager.getFileByPath(OCFile.ROOT_PATH);
        } else {
            safeInitialFolder = getFile();
        }

        MainFileListFragment mainListOfFiles = MainFileListFragment.newInstance(safeInitialFolder, true, FileListOption.ALL_FILES);
        mainListOfFiles.setFileActions(this);
        FragmentTransaction transaction = getSupportFragmentManager().beginTransaction();
        transaction.add(R.id.fragment_container, mainListOfFiles, TAG_LIST_OF_FOLDERS);
        transaction.commit();
    }

    private void setActionButtonText() {
        PickerMode actionButton = (PickerMode) getIntent().getSerializableExtra(EXTRA_PICKER_OPTION);
        Button chooseButton = findViewById(R.id.folder_picker_btn_choose);
        chooseButton.setText(getString(actionButton.getButtonString()));
    }

    protected MainFileListFragment getListOfFilesFragment() {
        Fragment listOfFiles = getSupportFragmentManager().findFragmentByTag(FolderPickerActivity.TAG_LIST_OF_FOLDERS);
        if (listOfFiles != null) {
            return (MainFileListFragment) listOfFiles;
        }
        Timber.e("Access to unexisting list of files fragment!!");
        return null;
    }

    @Override
    public void onSavedCertificate() {

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.main_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        boolean retval = true;
        switch (item.getItemId()) {
            case android.R.id.home: {
                OCFile currentDir = getCurrentFolder();
                if (currentDir != null && currentDir.getParentId() != 0) {
                    onBackPressed();
                }
                break;
            }
            default:
                retval = super.onOptionsItemSelected(item);
        }
        return retval;
    }

    protected OCFile getCurrentFolder() {
        MainFileListFragment listOfFiles = getListOfFilesFragment();
        if (listOfFiles != null) {  // should never be null, indeed
            return listOfFiles.getCurrentFile();
        }
        return null;
    }

    @Override
    public void onBackPressed() {
        MainFileListFragment listOfFiles = getListOfFilesFragment();
        if (listOfFiles != null) {  // should never be null, indeed
            OCFile fileBeforeBrowsingUp = listOfFiles.getCurrentFile();
            if (fileBeforeBrowsingUp.getParentId() != null &&
                    fileBeforeBrowsingUp.getParentId() == OCFile.ROOT_PARENT_ID
            ) {
                // If we are already at root, let's finish the picker. No sense to keep browsing up.
                finish();
                return;
            }
            listOfFiles.onBrowseUp();
            setFile(listOfFiles.getCurrentFile());
            updateNavigationElementsInActionBar();
        }
    }

    protected void updateNavigationElementsInActionBar() {
        OCFile currentDir;

        try {
            currentDir = getCurrentFolder();
        } catch (NullPointerException e) {
            currentDir = getFile();
        }

        boolean atRoot = (currentDir == null || currentDir.getParentId() == 0);
        updateStandardToolbar(
                atRoot ? getString(R.string.default_display_name_for_root_folder) : currentDir.getFileName(),
                !atRoot,
                !atRoot
        );
    }

    /**
     * Set per-view controllers
     */
    private void initPickerListeners() {
        mCancelBtn = findViewById(R.id.folder_picker_btn_cancel);
        mCancelBtn.setOnClickListener(this);
        mChooseBtn = findViewById(R.id.folder_picker_btn_choose);
        mChooseBtn.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        if (v == mCancelBtn) {
            finish();
        } else if (v == mChooseBtn) {
            Intent i = getIntent();
            ArrayList<Parcelable> targetFiles = i.getParcelableArrayListExtra(FolderPickerActivity.EXTRA_FILES);

            Intent data = new Intent();
            data.putExtra(EXTRA_FOLDER, getCurrentFolder());
            data.putParcelableArrayListExtra(EXTRA_FILES, targetFiles);
            setResult(RESULT_OK, data);

            finish();
        }
    }

    @Override
    public void onCurrentFolderUpdated(@NonNull OCFile newCurrentFolder, @Nullable OCSpace currentSpace) {
        updateNavigationElementsInActionBar();
        setFile(newCurrentFolder);
    }

    @Override
    public void initDownloadForSending(@NonNull OCFile file) {

    }

    @Override
    public void cancelFileTransference(@NonNull ArrayList<OCFile> files) {

    }

    @Override
    public void setBottomBarVisibility(boolean isVisible) {

    }

    @Override
    public void onFileClicked(@NonNull OCFile file) {
        // Nothing to do. Clicking on files is not allowed.
    }

    @Override
    public void onShareFileClicked(@NonNull OCFile file) {
        // Nothing to do. Clicking on files is not allowed.
    }

    @Override
    public void syncFile(@NonNull OCFile file) {
        // Nothing to do. Clicking on files is not allowed.
    }

    @Override
    public void openFile(@NonNull OCFile file) {
        // Nothing to do. Clicking on files is not allowed.
    }

    @Override
    public void sendDownloadedFile(@NonNull OCFile file) {
        // Nothing to do. Clicking on files is not allowed.

    }

    /**
     * Nothing to do. Details can't be opened from {@link FolderPickerActivity}
     *
     * @param file {@link OCFile} whose details will be shown
     */
    @Override
    public void showDetails(OCFile file) {

    }

    public enum PickerMode {
        MOVE, COPY, CAMERA_FOLDER;

        public Integer getButtonString() {
            switch (this) {
                case MOVE:
                    return R.string.folder_picker_move_here_button_text;
                case COPY:
                    return R.string.folder_picker_copy_here_button_text;
                default:
                    return R.string.folder_picker_choose_button_text;
            }
        }
    }
}

File: owncloudApp/src/main/java/com/owncloud/android/ui/activity/FolderPickerActivity.kt
/**
 * ownCloud Android client application
 *
 * @author Shashvat Kedia
 * @author David González Verdugo
 * @author Abel García de Prada
 * @author Juan Carlos Garrote Gascón
 * @author Aitor Ballesteros Pavón
 *
 * Copyright (C) 2023 ownCloud GmbH.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.owncloud.android.ui.activity

import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.annotation.StringRes
import androidx.core.view.isVisible
import com.owncloud.android.R
import com.owncloud.android.databinding.FilesFolderPickerBinding
import com.owncloud.android.datamodel.FileDataStorageManager
import com.owncloud.android.domain.files.model.FileListOption
import com.owncloud.android.domain.files.model.OCFile
import com.owncloud.android.domain.spaces.model.OCSpace
import com.owncloud.android.presentation.authentication.AccountUtils
import com.owncloud.android.presentation.files.filelist.MainFileListFragment
import com.owncloud.android.presentation.spaces.SpacesListFragment
import com.owncloud.android.ui.fragment.FileFragment
import com.owncloud.android.utils.PreferenceUtils
import timber.log.Timber

open class FolderPickerActivity : FileActivity(),
    FileFragment.ContainerActivity,
    MainFileListFragment.FileActions {

    protected val mainFileListFragment: MainFileListFragment?
        get() = supportFragmentManager.findFragmentByTag(TAG_LIST_OF_FOLDERS) as MainFileListFragment?

    private lateinit var pickerMode: PickerMode

    private lateinit var binding: FilesFolderPickerBinding



    override fun onCreate(savedInstanceState: Bundle?) {
        Timber.d("onCreate() start")

        super.onCreate(savedInstanceState)

        binding = FilesFolderPickerBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // Allow or disallow touches with other visible windows
        binding.filesFolderPickerLayout.filterTouchesWhenObscured = PreferenceUtils.shouldDisallowTouchesWithOtherVisibleWindows(this)

        pickerMode = intent.getSerializableExtra(EXTRA_PICKER_MODE) as PickerMode

        if (savedInstanceState == null) {
            when (pickerMode) {
                PickerMode.MOVE -> {
                    // Show the space where the files come from
                    val targetFiles = intent.getParcelableArrayListExtra<OCFile>(EXTRA_FILES)
                    val spaceIdOfFiles = targetFiles?.get(0)?.spaceId
                    initAndShowListOfFilesFragment(spaceId = spaceIdOfFiles)
                }
                PickerMode.COPY -> {
                    val targetFiles = intent.getParcelableArrayListExtra<OCFile>(EXTRA_FILES)
                    if (targetFiles?.get(0)?.spaceId != null) {
                        // Show the list of spaces
                        initAndShowListOfSpaces()
                    } else {
                        // Show the personal space
                        initAndShowListOfFilesFragment(spaceId = null)
                    }
                }
                PickerMode.CAMERA_FOLDER -> {
                    val spaceId = intent.getStringExtra(KEY_SPACE_ID)

                    if (spaceId != null) {
                        // Show the list of spaces
                        initAndShowListOfSpaces()
                    } else {
                        val accountName = intent.getStringExtra(KEY_ACCOUNT_NAME)
                        account = AccountUtils.getOwnCloudAccountByName(this, accountName)
                        // Show the personal space
                        initAndShowListOfFilesFragment(spaceId = null)
                    }
                }
            }
        }

        // Set callback listeners for UI elements
        initPickerListeners()

        // Action bar setup
        setupStandardToolbar(
            title = null,
            displayHomeAsUpEnabled = false,
            homeButtonEnabled = false,
            displayShowTitleEnabled = true,
        )

        // Set action button text
        setActionButtonText()

        supportFragmentManager.setFragmentResultListener(SpacesListFragment.REQUEST_KEY_CLICK_SPACE, this) { _, bundle ->
            val rootSpaceFolder = bundle.getParcelable<OCFile>(SpacesListFragment.BUNDLE_KEY_CLICK_SPACE)
            file = rootSpaceFolder
            initAndShowListOfFilesFragment()
        }

        Timber.d("onCreate() end")
    }

    override fun onResume() {
        super.onResume()
        updateToolbar(null, mainFileListFragment?.getCurrentSpace())
    }

    /**
     * Called when the ownCloud {@link Account} associated to the Activity was just updated.
     */
    override fun onAccountSet(stateWasRecovered: Boolean) {
        super.onAccountSet(stateWasRecovered)

        if (account != null) {
            updateFileFromDB()

            var folder = file
            if (folder == null || !folder.isFolder) {
                // Fall back to root folder
                file = storageManager.getRootPersonalFolder()
                folder = file
            }

            if (!stateWasRecovered) {
                mainFileListFragment?.navigateToFolder(folder)
            }

            updateNavigationElementsInActionBar()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        menu.removeItem(menu.findItem(R.id.action_share_current_folder)?.itemId ?: 0)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        val currentDirDisplayed = mainFileListFragment?.getCurrentFile()
        // If current file is null (we are in the spaces list, for example), close the activity
        if (currentDirDisplayed == null) {
            finish()
            return
        }
        // If current file is root folder
        else if (currentDirDisplayed.parentId == OCFile.ROOT_PARENT_ID) {
            // If we are not in COPY mode, or if we are in COPY mode and spaces are not allowed, close the activity
            if (pickerMode != PickerMode.COPY || (pickerMode == PickerMode.COPY && currentDirDisplayed.spaceId == null)) {
                finish()
                return
            }
            // If we are in COPY mode and inside a space, navigate back to the spaces list
            if (mainFileListFragment?.getCurrentSpace()?.isProject == true || mainFileListFragment?.getCurrentSpace()?.isPersonal == true) {
                file = null
                initAndShowListOfSpaces()
                updateToolbar(null)
                binding.folderPickerNoPermissionsMessage.isVisible = false
            }
        } else {
            mainFileListFragment?.onBrowseUp()
        }
    }

    override fun onCurrentFolderUpdated(newCurrentFolder: OCFile, currentSpace: OCSpace?) {
        updateToolbar(newCurrentFolder, currentSpace)
        updateButtonsVisibilityAccordingToPermissions(newCurrentFolder)
        file = newCurrentFolder
    }

    override fun initDownloadForSending(file: OCFile) {
        // Nothing to do. Downloading files is not allowed.
    }

    override fun cancelFileTransference(files: ArrayList<OCFile>) {
        // Nothing to do. Transferring files is not allowed.
    }

    override fun setBottomBarVisibility(isVisible: Boolean) {
        // Nothing to do. No changes will be done in the bottom bar visibility.
    }

    override fun onFileClicked(file: OCFile) {
        // Nothing to do. Clicking on files is not allowed.
    }

    override fun onShareFileClicked(file: OCFile) {
        // Nothing to do. Clicking on files is not allowed.
    }

    override fun syncFile(file: OCFile) {
        // Nothing to do. Clicking on files is not allowed.
    }

    override fun openFile(file: OCFile) {
        // Nothing to do. Clicking on files is not allowed.
    }

    override fun sendDownloadedFile(file: OCFile) {
        // Nothing to do. Clicking on files is not allowed.
    }

    override fun showDetails(file: OCFile) {
        // Nothing to do. Details can't be opened here.
    }

    private fun initAndShowListOfFilesFragment(spaceId: String? = null) {
        val safeInitialFolder = if (file == null) {
            if (account == null) {
                account = AccountUtils.getCurrentOwnCloudAccount(applicationContext)
            }
            val fileDataStorageManager = FileDataStorageManager(account)
            fileDataStorageManager.getFileByPath(OCFile.ROOT_PATH, spaceId)
        } else {
            file
        }

        file = safeInitialFolder

        safeInitialFolder?.let {
            val mainListOfFiles = MainFileListFragment.newInstance(it, true, FileListOption.ALL_FILES)
            mainListOfFiles.fileActions = this
            val transaction = supportFragmentManager.beginTransaction()
            transaction.replace(R.id.fragment_container, mainListOfFiles, TAG_LIST_OF_FOLDERS)
            transaction.commit()
        }


        binding.folderPickerBtnChoose.isVisible = true
    }

    private fun initAndShowListOfSpaces() {
        val accountNameIntent = intent.getStringExtra(KEY_ACCOUNT_NAME)
        val accountName = accountNameIntent ?: AccountUtils.getCurrentOwnCloudAccount(applicationContext).name

        val listOfSpaces = SpacesListFragment.newInstance(showPersonalSpace = true, accountName = accountName)
        val transaction = supportFragmentManager.beginTransaction()
        transaction.replace(R.id.fragment_container, listOfSpaces)
        transaction.commit()
        binding.folderPickerBtnChoose.isVisible = false
    }

    /**
     * Set per-view controllers
     */
    private fun initPickerListeners() {

        binding.folderPickerBtnCancel.setOnClickListener {
            finish()
        }

        binding.folderPickerBtnChoose.setOnClickListener {
            val data = Intent().apply {
                val targetFiles = intent.getParcelableArrayListExtra<OCFile>(EXTRA_FILES)
                putExtra(EXTRA_FOLDER, getCurrentFolder())
                putParcelableArrayListExtra(EXTRA_FILES, targetFiles)
            }
            setResult(RESULT_OK, data)

            finish()
        }
    }

    private fun setActionButtonText() {
        binding.folderPickerBtnChoose.text = getString(pickerMode.toStringRes())
    }

    private fun getCurrentFolder(): OCFile? {
        if (mainFileListFragment != null) {
            return mainFileListFragment?.getCurrentFile()
        }
        return null
    }

    private fun updateToolbar(chosenFileFromParam: OCFile?, space: OCSpace? = null) {
        val chosenFile = chosenFileFromParam ?: file // If no file is passed, current file decides
        val isRootFromPersonalInCopyMode =
            chosenFile != null && chosenFile.remotePath == OCFile.ROOT_PATH && space?.isProject == false && pickerMode == PickerMode.COPY
        val isRootFromPersonal = chosenFile == null || (chosenFile.remotePath == OCFile.ROOT_PATH && (space == null || !space.isProject))
        val isRootFromProject = space?.isProject == true && chosenFile.remotePath == OCFile.ROOT_PATH

        if (isRootFromPersonalInCopyMode) {
            updateStandardToolbar(
                title = getString(R.string.default_display_name_for_root_folder),
                displayHomeAsUpEnabled = true,
                homeButtonEnabled = true
            )
        } else if (isRootFromPersonal) {
            updateStandardToolbar(
                title = getString(R.string.default_display_name_for_root_folder),
                displayHomeAsUpEnabled = false,
                homeButtonEnabled = false
            )
        } else if (isRootFromProject) {
            updateStandardToolbar(
                title = space!!.name,
                displayHomeAsUpEnabled = pickerMode == PickerMode.COPY,
                homeButtonEnabled = pickerMode == PickerMode.COPY
            )
        } else {
            updateStandardToolbar(title = chosenFile.fileName, displayHomeAsUpEnabled = true, homeButtonEnabled = true)
        }
    }

    private fun updateButtonsVisibilityAccordingToPermissions(currentFolder: OCFile) {
        currentFolder.hasAddFilePermission.let {
            binding.folderPickerBtnChoose.isVisible = it
            binding.folderPickerNoPermissionsMessage.isVisible = !it
        }
    }

    protected fun updateNavigationElementsInActionBar() {
        val currentDir = try {
            getCurrentFolder()
        } catch (e: NullPointerException) {
            file
        }

        val atRoot = (currentDir == null || currentDir.parentId == 0L)
        updateStandardToolbar(
            title = if (atRoot) getString(R.string.default_display_name_for_root_folder) else currentDir!!.fileName,
            displayHomeAsUpEnabled = !atRoot,
            homeButtonEnabled = !atRoot,
        )
    }

    enum class PickerMode {
        MOVE, COPY, CAMERA_FOLDER;

        @StringRes
        fun toStringRes(): Int {
            return when (this) {
                MOVE -> R.string.folder_picker_move_here_button_text
                COPY -> R.string.folder_picker_copy_here_button_text
                CAMERA_FOLDER -> R.string.folder_picker_choose_button_text
            }
        }
    }

    companion object {
        const val KEY_ACCOUNT_NAME = "KEY_ACCOUNT_NAME"
        const val KEY_SPACE_ID = "KEY_PERSONAL_SPACE_ID"
        const val EXTRA_FOLDER = "FOLDER_PICKER_EXTRA_FOLDER"
        const val EXTRA_FILES = "FOLDER_PICKER_EXTRA_FILES"
        const val EXTRA_PICKER_MODE = "FOLDER_PICKER_EXTRA_PICKER_MODE"
        private const val TAG_LIST_OF_FOLDERS = "LIST_OF_FOLDERS"
    }
}

